src/daft-io/src/s3_like.rs

```diff
@@ -227,13 +263,21 @@ impl From<Error> for super::Error {
                         path,
                         source: no_such_key.into(),
                     },
-                    ListObjectsV2Error::Unhandled(v) => super::Error::Unhandled {
-                        path,
-                        msg: DisplayErrorContext(v).to_string(),
-                    },
-                    err => super::Error::UnableToOpenFile {
-                        path,
-                        source: err.into(),
+                    err => match err.code() {
+                        Some("InternalError") => super::Error::MiscTransient {
+                            path,
+                            source: err.into(),
+                        },
+                        Some(code) if THROTTLING_ERRORS.contains(&code) => {
+                            super::Error::Throttled {
+                                path,
+                                source: err.into(),
+                            }
+                        }
+                        _ => super::Error::Unhandled {
+                            path,
+                            msg: DisplayErrorContext(err).to_string(),
+                        },
                     },
```

Comment:
Seems like this group of logic is being repeated in 3 different places. Any chance that we could extract it into its own function?

Reply:
Done.